awslabs/sagemaker-debugger-rulesconfig
# Add actions for debugger rules and set up PR CI for unit tests

*Issue #, if available:*

*Description of changes:*
Add action classes to be used by the sagemaker SDK to simplify the usage of actions to be invoked when debugger rules fire during training.

Tested these changes by:
- Building a rules config wheel with these changes
- Building a sagemaker SDK wheel with the changes in [this branch](https://github.com/ndodda-amazon/sagemaker-python-sdk/tree/actions).
- Using a training script that sleeps for 20 minutes after training (before the script ends)
- Running a training job with stalled training rule with threshold 15 minutes and all three of the new actions.
- Confirmed that the stalled training rule fired, the training job was stopped early, and both an email/SMS was sent to the email address/phone number specified in the actions.

See code block below for user experience:

```
actions = rule_configs.ActionList(rule_configs.StopTraining(), rule_configs.Email("abc@abc.com"), rule_configs.SMS("+1234567890"))
rules = [
    Rule.sagemaker(rule_configs.vanishing_gradient()), 
    Rule.sagemaker(rule_configs.loss_not_decreasing()),
    Rule.sagemaker(
        rule_configs.stalled_training_rule(), 
        rule_parameters={"threshold": str(60*15)},
        actions=actions
    )
]


estimator = TensorFlow(
    role=sagemaker.get_execution_role(),
    base_job_name='test-actions',
    instance_count=1,
    instance_type='ml.c5.4xlarge',
    framework_version='1.15',
    py_version='py3',
    source_dir="src",
    entry_point="simple_stalled_training.py",
    rules = rules
)
```

Also added unit tests for these changes and set up the infrastructure to run these unit tests as a part of the PR CI. Since the codebase itself is framework independent, we can have a separate build for just running unit tests (and any future builds would only run sagemaker tests for the desired framework + version).

TODO:
- Add end to end sagemaker tests for these changes
- Set up code coverage

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
